### ISSUE TITLE
Locking issues when uploading several files

### ISSUE THREAD

**Issue by ojlyytinen** [projecthydra/sufia#1362]
Although this issue is similar to #1324, it is different in that it has nothing to do with uploading to a collection. The cause seems to be essentially trying to create the Batch object with the same ID twice when uploading several files. It appears that the code [Batch.safe_create](https://github.com/projecthydra/sufia/blob/master/sufia-models/app/models/batch.rb#L22-L29) doesn't sufficiently handle the race condition.

This doesn't happen every single time you upload several files but we ran into it twice in a fairly short timespan. Below are the two stack dumps from our logs. They are similar but not exactly the same. But it seems like the cause is still the same, the conflict is just detected at different points of execution.

```
E, [2015-11-24T15:13:53.480719 #17721] ERROR -- : GenericFilesController::create rescued Ldp::HttpError
	STATUS: 500 javax.jcr.ItemExistsException: A node definition that allows same name siblings could not be found for the node "{}42" in workspace "default"
	at org.modeshape.jcr.JcrSession$JcrPreSave.processAfterLocking(JcrSession.java:2422)
	at org.modeshape.jcr.cache.document.WritableSessionCache.runPreSaveAfterLocking(WritableSessionCache.java:541)
	at org.modeshape.jcr.cache.document.WritableSessionCache.save(WritableSessionCache.java:614)
	at org.modeshape.jcr.JcrSession.save(JcrSession.java:1166)
	at org.fcrepo.http.api.FedoraLdp.createOrReplaceObjectRdf(FedoraLdp.java:301)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at org.glassfish.jersey.server.model.internal.ResourceMethodInvocationHandlerFactory$1.invoke(ResourceMethodInvocationHandlerFactory.java...
 /var/www/hydra/vendor/bundle/ruby/gems/ldp-0.4.1/lib/ldp/client/methods.rb:121:in `block in check_for_errors'
/var/www/hydra/vendor/bundle/ruby/gems/ldp-0.4.1/lib/ldp/client/methods.rb:119:in `tap'
/var/www/hydra/vendor/bundle/ruby/gems/ldp-0.4.1/lib/ldp/client/methods.rb:119:in `check_for_errors'
/var/www/hydra/vendor/bundle/ruby/gems/ldp-0.4.1/lib/ldp/client/methods.rb:98:in `block in put'
/var/www/hydra/vendor/bundle/ruby/gems/activesupport-4.2.1/lib/active_support/notifications.rb:166:in `instrument'
/var/www/hydra/vendor/bundle/ruby/gems/ldp-0.4.1/lib/ldp/client/methods.rb:90:in `put'
/var/www/hydra/vendor/bundle/ruby/gems/active-fedora-9.4.2/lib/active_fedora/caching_connection.rb:24:in `put'
/var/www/hydra/vendor/bundle/ruby/gems/ldp-0.4.1/lib/ldp/resource.rb:72:in `create'
/var/www/hydra/vendor/bundle/ruby/gems/ldp-0.4.1/lib/ldp/resource/rdf_source.rb:21:in `create'
/var/www/hydra/vendor/bundle/ruby/gems/active-fedora-9.4.2/lib/active_fedora/persistence.rb:147:in `create_record'
/var/www/hydra/vendor/bundle/ruby/gems/active-fedora-9.4.2/lib/active_fedora/indexing.rb:53:in `create_record'
/var/www/hydra/vendor/bundle/ruby/gems/active-fedora-9.4.2/lib/active_fedora/callbacks.rb:237:in `block (2 levels) in create_record'
/var/www/hydra/vendor/bundle/ruby/gems/activesupport-4.2.1/lib/active_support/callbacks.rb:117:in `call'
/var/www/hydra/vendor/bundle/ruby/gems/activesupport-4.2.1/lib/active_support/callbacks.rb:117:in `call'
/var/www/hydra/vendor/bundle/ruby/gems/activesupport-4.2.1/lib/active_support/callbacks.rb:555:in `block (2 levels) in compile'
/var/www/hydra/vendor/bundle/ruby/gems/activesupport-4.2.1/lib/active_support/callbacks.rb:505:in `call'
/var/www/hydra/vendor/bundle/ruby/gems/activesupport-4.2.1/lib/active_support/callbacks.rb:505:in `call'
/var/www/hydra/vendor/bundle/ruby/gems/activesupport-4.2.1/lib/active_support/callbacks.rb:92:in `_run_callbacks'
/var/www/hydra/vendor/bundle/ruby/gems/activesupport-4.2.1/lib/active_support/callbacks.rb:776:in `_run_save_callbacks'
/var/www/hydra/vendor/bundle/ruby/gems/activesupport-4.2.1/lib/active_support/callbacks.rb:81:in `run_callbacks'
/var/www/hydra/vendor/bundle/ruby/gems/active-fedora-9.4.2/lib/active_fedora/callbacks.rb:237:in `block in create_record'
/var/www/hydra/vendor/bundle/ruby/gems/activesupport-4.2.1/lib/active_support/callbacks.rb:117:in `call'
/var/www/hydra/vendor/bundle/ruby/gems/activesupport-4.2.1/lib/active_support/callbacks.rb:117:in `call'
/var/www/hydra/vendor/bundle/ruby/gems/activesupport-4.2.1/lib/active_support/callbacks.rb:555:in `block (2 levels) in compile'
/var/www/hydra/vendor/bundle/ruby/gems/activesupport-4.2.1/lib/active_support/callbacks.rb:505:in `call'
/var/www/hydra/vendor/bundle/ruby/gems/activesupport-4.2.1/lib/active_support/callbacks.rb:505:in `call'
/var/www/hydra/vendor/bundle/ruby/gems/activesupport-4.2.1/lib/active_support/callbacks.rb:92:in `_run_callbacks'
/var/www/hydra/vendor/bundle/ruby/gems/activesupport-4.2.1/lib/active_support/callbacks.rb:776:in `_run_create_callbacks'
/var/www/hydra/vendor/bundle/ruby/gems/activesupport-4.2.1/lib/active_support/callbacks.rb:81:in `run_callbacks'
/var/www/hydra/vendor/bundle/ruby/gems/active-fedora-9.4.2/lib/active_fedora/callbacks.rb:236:in `create_record'
/var/www/hydra/vendor/bundle/ruby/gems/active-fedora-9.4.2/lib/active_fedora/persistence.rb:139:in `create_or_update'
/var/www/hydra/vendor/bundle/ruby/gems/active-fedora-9.4.2/lib/active_fedora/persistence.rb:27:in `save'
/var/www/hydra/vendor/bundle/ruby/gems/active-fedora-9.4.2/lib/active_fedora/validations.rb:50:in `save'
/var/www/hydra/vendor/bundle/ruby/gems/active-fedora-9.4.2/lib/active_fedora/persistence.rb:107:in `create'
/var/www/hydra/vendor/bundle/ruby/bundler/gems/sufia-f8387043b826/sufia-models/app/models/batch.rb:23:in `safe_create'
/var/www/hydra/vendor/bundle/ruby/bundler/gems/sufia-f8387043b826/sufia-models/app/models/batch.rb:15:in `rescue in find_or_create'
/var/www/hydra/vendor/bundle/ruby/bundler/gems/sufia-f8387043b826/sufia-models/app/models/batch.rb:13:in `find_or_create'
/var/www/hydra/vendor/bundle/ruby/bundler/gems/sufia-f8387043b826/app/controllers/concerns/sufia/files_controller_behavior.rb:236:in `process_file'
/var/www/hydra/vendor/bundle/ruby/bundler/gems/sufia-f8387043b826/app/controllers/concerns/sufia/files_controller_behavior.rb:105:in `create_from_upload'
/var/www/hydra/vendor/bundle/ruby/bundler/gems/sufia-f8387043b826/app/controllers/concerns/sufia/files_controller_behavior.rb:90:in `create'
/var/www/hydra/vendor/bundle/ruby/bundler/gems/sufia-f8387043b826/app/controllers/concerns/sufia/files_controller/browse_everything.rb:9:in `create'
/var/www/hydra/vendor/bundle/ruby/bundler/gems/sufia-f8387043b826/app/controllers/concerns/sufia/files_controller/local_ingest_behavior.rb:9:in `create'
/var/www/hydra/vendor/bundle/ruby/gems/actionpack-4.2.1/lib/action_controller/metal/implicit_render.rb:4:in `send_action'
/var/www/hydra/vendor/bundle/ruby/gems/actionpack-4.2.1/lib/abstract_controller/base.rb:198:in `process_action'

...
```

```
E, [2015-11-24T15:17:26.404609 #7864] ERROR -- : GenericFilesController::create rescued RuntimeError
	Can't call create on an existing resource
 /var/www/hydra/vendor/bundle/ruby/gems/ldp-0.4.1/lib/ldp/resource.rb:70:in `create'
/var/www/hydra/vendor/bundle/ruby/gems/ldp-0.4.1/lib/ldp/resource/rdf_source.rb:21:in `create'
/var/www/hydra/vendor/bundle/ruby/gems/active-fedora-9.4.2/lib/active_fedora/persistence.rb:147:in `create_record'
/var/www/hydra/vendor/bundle/ruby/gems/active-fedora-9.4.2/lib/active_fedora/indexing.rb:53:in `create_record'
/var/www/hydra/vendor/bundle/ruby/gems/active-fedora-9.4.2/lib/active_fedora/callbacks.rb:237:in `block (2 levels) in create_record'
/var/www/hydra/vendor/bundle/ruby/gems/activesupport-4.2.1/lib/active_support/callbacks.rb:117:in `call'
/var/www/hydra/vendor/bundle/ruby/gems/activesupport-4.2.1/lib/active_support/callbacks.rb:117:in `call'
/var/www/hydra/vendor/bundle/ruby/gems/activesupport-4.2.1/lib/active_support/callbacks.rb:555:in `block (2 levels) in compile'
/var/www/hydra/vendor/bundle/ruby/gems/activesupport-4.2.1/lib/active_support/callbacks.rb:505:in `call'
/var/www/hydra/vendor/bundle/ruby/gems/activesupport-4.2.1/lib/active_support/callbacks.rb:505:in `call'
/var/www/hydra/vendor/bundle/ruby/gems/activesupport-4.2.1/lib/active_support/callbacks.rb:92:in `_run_callbacks'
/var/www/hydra/vendor/bundle/ruby/gems/activesupport-4.2.1/lib/active_support/callbacks.rb:776:in `_run_save_callbacks'
/var/www/hydra/vendor/bundle/ruby/gems/activesupport-4.2.1/lib/active_support/callbacks.rb:81:in `run_callbacks'
/var/www/hydra/vendor/bundle/ruby/gems/active-fedora-9.4.2/lib/active_fedora/callbacks.rb:237:in `block in create_record'
/var/www/hydra/vendor/bundle/ruby/gems/activesupport-4.2.1/lib/active_support/callbacks.rb:117:in `call'
/var/www/hydra/vendor/bundle/ruby/gems/activesupport-4.2.1/lib/active_support/callbacks.rb:117:in `call'
/var/www/hydra/vendor/bundle/ruby/gems/activesupport-4.2.1/lib/active_support/callbacks.rb:555:in `block (2 levels) in compile'
/var/www/hydra/vendor/bundle/ruby/gems/activesupport-4.2.1/lib/active_support/callbacks.rb:505:in `call'
/var/www/hydra/vendor/bundle/ruby/gems/activesupport-4.2.1/lib/active_support/callbacks.rb:505:in `call'
/var/www/hydra/vendor/bundle/ruby/gems/activesupport-4.2.1/lib/active_support/callbacks.rb:92:in `_run_callbacks'
/var/www/hydra/vendor/bundle/ruby/gems/activesupport-4.2.1/lib/active_support/callbacks.rb:776:in `_run_create_callbacks'
/var/www/hydra/vendor/bundle/ruby/gems/activesupport-4.2.1/lib/active_support/callbacks.rb:81:in `run_callbacks'
/var/www/hydra/vendor/bundle/ruby/gems/active-fedora-9.4.2/lib/active_fedora/callbacks.rb:236:in `create_record'
/var/www/hydra/vendor/bundle/ruby/gems/active-fedora-9.4.2/lib/active_fedora/persistence.rb:139:in `create_or_update'
/var/www/hydra/vendor/bundle/ruby/gems/active-fedora-9.4.2/lib/active_fedora/persistence.rb:27:in `save'
/var/www/hydra/vendor/bundle/ruby/gems/active-fedora-9.4.2/lib/active_fedora/validations.rb:50:in `save'
/var/www/hydra/vendor/bundle/ruby/gems/active-fedora-9.4.2/lib/active_fedora/persistence.rb:107:in `create'
/var/www/hydra/vendor/bundle/ruby/bundler/gems/sufia-f8387043b826/sufia-models/app/models/batch.rb:23:in `safe_create'
/var/www/hydra/vendor/bundle/ruby/bundler/gems/sufia-f8387043b826/sufia-models/app/models/batch.rb:15:in `rescue in find_or_create'
/var/www/hydra/vendor/bundle/ruby/bundler/gems/sufia-f8387043b826/sufia-models/app/models/batch.rb:13:in `find_or_create'
/var/www/hydra/vendor/bundle/ruby/bundler/gems/sufia-f8387043b826/app/controllers/concerns/sufia/files_controller_behavior.rb:236:in `process_file'
/var/www/hydra/vendor/bundle/ruby/bundler/gems/sufia-f8387043b826/app/controllers/concerns/sufia/files_controller_behavior.rb:105:in `create_from_upload'
/var/www/hydra/vendor/bundle/ruby/bundler/gems/sufia-f8387043b826/app/controllers/concerns/sufia/files_controller_behavior.rb:90:in `create'
/var/www/hydra/vendor/bundle/ruby/bundler/gems/sufia-f8387043b826/app/controllers/concerns/sufia/files_controller/browse_everything.rb:9:in `create'
/var/www/hydra/vendor/bundle/ruby/bundler/gems/sufia-f8387043b826/app/controllers/concerns/sufia/files_controller/local_ingest_behavior.rb:9:in `create'
/var/www/hydra/vendor/bundle/ruby/gems/actionpack-4.2.1/lib/action_controller/metal/implicit_render.rb:4:in `send_action'
/var/www/hydra/vendor/bundle/ruby/gems/actionpack-4.2.1/lib/abstract_controller/base.rb:198:in `process_action'

...
```

**Comment by jcoyne** [comment]
Yes, Batch.safe_create is too naive. It should be wrapped with `Sufia::LockManager#lock`

**Comment by mjgiarlo** [comment]
This issue was moved to projecthydra-labs/curation_concerns#560

**Comment by hackmastera** [comment]
Moving this issue implies that any solution won't be backported to 6.x? Please confirm.

**Comment by mjgiarlo** [comment]
@hackmastera I wouldn't say this means that a solution *won't* be backported. I can take a look at this today.

**Comment by mjgiarlo** [comment]
Reopening for Sufia 6.x

**Comment by mjgiarlo** [comment]
Fixed by #1484
